@xen-orchestra/fs/src/abstract.js

```diff
@@ -118,56 +118,48 @@ export default class RemoteHandlerAbstract {
     await this.__closeFile(fd)
   }
 
-  createReadStream(file, { checksum = false, ignoreMissingChecksum = false, ...options } = {}) {
+  async createReadStream(file, { checksum = false, ignoreMissingChecksum = false, ...options } = {}) {
     if (typeof file === 'string') {
       file = normalizePath(file)
     }
-    const path = typeof file === 'string' ? file : file.path
-    const streamP = timeout
-      .call(this._createReadStream(file, { ...options, highWaterMark: this._highWaterMark }), this._timeout)
-      .then(stream => {
-        // detect early errors
-        let promise = fromEvent(stream, 'readable')
-
-        // try to add the length prop if missing and not a range stream
-        if (stream.length === undefined && options.end === undefined && options.start === undefined) {
-          promise = Promise.all([
-            promise,
-            ignoreErrors.call(
-              this._getSize(file).then(size => {
-                stream.length = size
-              })
-            ),
-          ])
-        }
 
-        return promise.then(() => stream)
-      })
+    let stream = await timeout.call(
+      this._createReadStream(file, { ...options, highWaterMark: this._highWaterMark }),
+      this._timeout
+    )
 
-    const decryptedStream = this._encryptor.decryptStream(streamP)
-    if (!checksum) {
-      return decryptedStream
+    // detect early errors
+    await fromEvent(stream, 'readable')
+
+    if (checksum) {
+      try {
+        const path = typeof file === 'string' ? file : file.path
+        const checksum = await this._readFile(checksumFile(path), { flags: 'r' })
+
+        const { length } = stream
+        stream = validChecksumOfReadStream(stream, String(checksum).trim())
+        stream.length = length
+      } catch (error) {
+        if (!(ignoreMissingChecksum && error.code === 'ENOENT')) {
+          throw error
+        }
+      }
     }
 
-    // avoid a unhandled rejection warning
-    ignoreErrors.call(decryptedStream)
-
-    return this._readFile(checksumFile(path), { flags: 'r' }).then(
-      checksum =>
-        decryptedStream.then(stream => {
-          const { length } = stream
-          stream = validChecksumOfReadStream(stream, String(checksum).trim())
-          stream.length = length
-
-          return stream
-        }),
-      error => {
-        if (ignoreMissingChecksum && error && error.code === 'ENOENT') {
-          return decryptedStream
+    if (this.isEncrypted) {
+      stream = this._encryptor.decryptStream(stream)
```

Comment:
start and end options seems to be  used only for tests, but I think we should throw an error when trying to use them with encrypted files

Reply:
That makes sense, can you take care of it?

Reply:
At the beginning of the function, no need to do anything in that case.